pkg/inframanager/p4/policy.go

```diff
@@ -0,0 +1,866 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build dpdk
+
+package p4
+
+import (
+	"context"
+	"fmt"
+	"github.com/antoninbas/p4runtime-go-client/pkg/client"
+	"github.com/ipdk-io/k8s-infra-offload/pkg/inframanager/store"
+	"strconv"
+	"strings"
+	//	p4_v1 "github.com/p4lang/p4runtime/go/p4/v1"
+	log "github.com/sirupsen/logrus"
+	//	"net"
+)
+
+type tabletype int
+
+const (
+	denyall tabletype = iota
+	policyadd
+	policydel
+	policyupdate
+	workloadadd
+	workloadupdate
+	workloaddel
+)
+
+func AclPodIpProtoTableEgress(ctx context.Context, p4RtC *client.Client,
+	protocol string, workerep string, polID uint16, rangeID uint16,
+	action InterfaceType) error {
+	switch action {
+	case Insert:
+		if protocol != "" {
+			entryAdd := p4RtC.NewTableEntry(
+				"k8s_dp_control.acl_pod_ip_proto_table_egress",
+				map[string]client.MatchInterface{
+					"hdr.ipv4.src_addr": &client.ExactMatch{
+						Value: Pack32BinaryIP4(workerep),
+					},
+					"hdr.ipv4.protocol": &client.LpmMatch{
+						Value: valueToBytesStr(protocol),
+						PLen:  16,
+					},
+				},
+				p4RtC.NewTableActionDirect("k8s_dp_control.set_range_check_ref",
+					[][]byte{valueToBytes16(polID),
+						valueToBytes16(rangeID)}),
+				nil,
+			)
+			if err := p4RtC.InsertTableEntry(ctx, entryAdd); err != nil {
+				log.Errorf("Cannot insert entry into 'acl_pod_ip_proto_table_egress': %v", err)
+				return err
+			}
+		} else {
+			entryAdd := p4RtC.NewTableEntry(
+				"k8s_dp_control.acl_pod_ip_proto_table_egress",
+				map[string]client.MatchInterface{
+					"hdr.ipv4.src_addr": &client.ExactMatch{
+						Value: Pack32BinaryIP4(workerep),
+					},
+					"hdr.ipv4.protocol": &client.LpmMatch{
+						Value: valueToBytesStr(protocol),
+						PLen:  16,
+					},
+				},
+				p4RtC.NewTableActionDirect("k8s_dp_control.set_status_match_ipset_only",
+					[][]byte{valueToBytes16(polID)}),
+				nil,
+			)
+			if err := p4RtC.InsertTableEntry(ctx, entryAdd); err != nil {
+				log.Errorf("Cannot insert entry into 'acl_pod_ip_proto_table_egress': %v", err)
+				return err
+			}
+		}
+
+	case Delete:
+		entryDelete := p4RtC.NewTableEntry(
+			"k8s_dp_control.acl_pod_ip_proto_table_egress",
+			map[string]client.MatchInterface{
+				"hdr.ipv4.src_addr": &client.ExactMatch{
+					Value: Pack32BinaryIP4(workerep),
+				},
+				"hdr.ipv4.protocol": &client.LpmMatch{
+					Value: valueToBytesStr(protocol),
+					PLen:  16,
+				},
+			},
+			nil,
+			nil,
+		)
+		if err := p4RtC.DeleteTableEntry(ctx, entryDelete); err != nil {
+			log.Errorf("Cannot delete entry from 'acl_pod_ip_proto_table_egress': %v", err)
+			return err
+		}
+
+	default:
+		log.Warnf("Invalid action %v", action)
+		err := fmt.Errorf("Invalid action %v", action)
+		return err
+	}
+
+	return nil
+}
+
+func AclPodIpProtoTableIngress(ctx context.Context, p4RtC *client.Client,
```

Comment:
Agree with having common functions to avoid duplicate code. Since the only differences in the two functions are table/action names, can we have string variables for these and decide the value based on the direction/protocol? Then we can have common code for Ingress/Egress and can avoid the if-else for protocol. This will greatly reduce redundant code.